### ISSUE TITLE
Audio and Video out of sync

### ISSUE THREAD

**Issue by jmasero** [microsoftgraph/microsoft-graph-comms-samples#514]
Hello!

We are writing a bot application using MS Teams SDK.
Basically we want to record audio and video that comes in H264 Variable Frame Rate format, but when receiving video frames (H264), the resulting file is out of sync with the related audio stream.
The timestamps are checked, but the video file is always with less frames than expected.

Our bot performs streams management by application side (local self-host server).
It's based on .NET FW and Microsft.Skype.Bots.Media library. The video streams are in H264 format.

However we get unsync audio/video due to the fact that Teams sends this video in variable frame rate. Is there a way to get a syncronized audio/video?

**Comment by 1fabi0** [comment]
You can use the timestamps that are sent with the video to do a proper synchronization with a custom logic, take the average framerate of the videos with the timestamps and play with the average framerate (probably then its  sometimes slightly out of sync) or you can try with ffmpeg pipe input and clock timings of the frames to replay with approximately correct timings. (First solution is probably the best to solve this issue)

**Comment by mosoftwareenterprises** [comment]
@jmasero did you solve this issue. I am trying to do it with ffmpeg. Any tips?

**Comment by jmasero** [comment]
Hello ... we have fixed the sync issues, using a H264 frame decoder and the ffmpeg to re-encode to MP4 format. The re-encode process needs the timestamps for all the video frames recorded by the bots media library.

All of those jobs are to get a constant-frame-rate video stream.

**Comment by 1fabi0** [comment]
@jmasero I don't see a reason to use a special decoder, there is a built in decoder in the media sdk to already decode to all kind of image formats like rgb24 or nv12 and much more, so you could have just used the built in decoder.

Or do I miss understand you're solution?

You are decoding the images and forward them with a constant framerate to ffmpeg to let ffmpeg create the mp4, right?

**Comment by jmasero** [comment]
Hi, the bot receives the video frames in H264 format, but in variable-frame-rate (it depends of the bandwith of the participants, the MS media servers workload). With that H264 VFR file, the video and audio can't be synchronized.

We need always a CFR video file to allow a perfect audio sync.

That is the reason to use the H264 decoder.

**Comment by mosoftwareenterprises** [comment]
I dont suppose you have a blog post or code to share (privately?)

**Comment by jmasero** [comment]
Hi, we are deploying in Azure Service Fabric ... at corporative project. We need to ensure a load balancing environment to take the recording bot solution as OK.

The source code is not available to share (at this moment), but the bot uses FFMPEG in some steps of conversion H264 --> MP4.

We are sorry about this.

**Comment by 1fabi0** [comment]
> Hi, the bot receives the video frames in H264 format, but in variable-frame-rate (it depends of the bandwith of the participants, the MS media servers workload). With that H264 VFR file, the video and audio can't be synchronized.

That's totally right but I want to enhance it with more info for education the media sdk of the bot can decode the h264 video to various formats, and at the point of calling the onreceived events audio and video is in sync, so if you would directly playout audio on receiving and show frames when they are received everything would be in sync, but as you have stated out the framerates are not constant.

As a small request here, maybe you have some block posts, documentation, something else that was helpful for your solution to convert timestamps and raw h264 files to a constant frame rate video, e.g. A link to a ffmpeg parameter that is helpful, a stack overflow answer that was helpful etc.
I think it would be great for everyone if we would have such sources here.

**Comment by jmasero** [comment]
The library KDAVCodec, was developed by **Krystian Dużyński** (https://github.com/KrystianD?tab=repositories).

**Comment by KrystianD** [comment]
@1fabi0 this process is much more complex than just using ffmpeg. Our first step was to adjust audio and video beginnings and durations, then, we use a custom tool (that internall uses ffmpeg library) for adjusting `pts` and `dts` of the h264 frames with timestamps saved from the bot. This resulted in fixed h264 file. Last step was to merge .wav file and fixed h264 file in the constant frame rate mp4 file with ffmpeg command.

**Comment by 1fabi0** [comment]
@KrystianD thanks a lot for that, I know that it is a complex process, but the info you provided is way more accurate and helpful than the stuff we had before.

**Comment by JonatanMedinilla** [comment]
@mosoftwareenterprises @1fabi0 I don't know if this will answer your questions, but a year ago I've been working on an [open source project](https://github.com/microsoft/Broadcast-Development-Kit) that relies on GStreamer to handle the media processing (decoding, encoding, etc).  

The code might not be the best, but in this sample, we use GStreamer to:
- decode the frames so we can fix the frame rate, frame size, and aspect ratio.
- synchronize the audio and the video with the timestamps provided by the media platform
- encode the audio and video to the corresponding formats that we want to use to stream the participant's feed via SRT or RTMP (e.g: h264 and aac for SRT/RTMP streaming).

I will recommend both to learn about GStreamer. If you want to record the participant's feed to a file instead of streaming it over SRT/RTMP, it is a manner of changing the corresponding plugins on the pipeline to record the media into a file (e.g.: _qtmux_ to mux/merge the h264 video and aac audio into .mov/.mp4 files, and _filesink_ to save the video.

**Comment by JonatanMedinilla** [comment]
@mosoftwareenterprises @1fabi0 I don't know if this will answer your questions, but a year ago I've been working on an [open source project](https://github.com/microsoft/Broadcast-Development-Kit) that relies on GStreamer to handle the media processing (decoding, encoding, etc).  

The code might not be the best, but in this sample we use GStreamer to:
- decode the frames so we can fix the frame rate, frame size, and aspect ratio.
- synchronize the audio and the video with the timestamps provided by the media platform
- encode the audio and video to the corresponding formats that we want to use to stream the participant's feed via SRT or RTMP (e.g: h264 and aac for SRT/RTMP streaming).

I will recommend you both to read about GStreamer. If you want to record the participant's feed to a file instead of SRT/RTMP streaming, it is a matter of changing the corresponding plugins in the pipeline to record the media to a file (for example _qtmux_ to merge the h264 video and aac audio into .mov/.mp4 files, and _filesink_ to save the video.
